File: 2025/09 September/13 Minimum Cost to cut a board into squares/main.py
# Link: https://www.geeksforgeeks.org/problems/minimum-cost-to-cut-a-board-into-squares/1


class Solution:
    def minCost(self, n, m, x, y):
        x.sort()
        y.sort()

        hCount, vCount = 1, 1
        i, j = len(x) - 1, len(y) - 1
        totalCost = 0
        while i >= 0 and j >= 0:
            if x[i] >= y[j]:
                totalCost += x[i] * hCount
                vCount += 1
                i -= 1
            else:
                totalCost += y[j] * vCount
                hCount += 1
                j -= 1

        while i >= 0:
            totalCost += x[i] * hCount
            vCount += 1
            i -= 1

        while j >= 0:
            totalCost += y[j] * vCount
            hCount += 1
            j -= 1

        return totalCost


if __name__ == "__main__":
    sol = Solution()
    n, m = 4, 6
    x = [2, 1, 3, 1, 4]
    y = [4, 1, 2]
    result = sol.minCost(n, m, x, y)
    print(f"Minimum cost to cut the board into squares: {result}")

    n, m = 4, 4
    x = [1, 1, 1]
    y = [1, 1, 1]
    result = sol.minCost(n, m, x, y)
    print(f"Minimum cost to cut the board into squares: {result}")

File: 2025/09 September/13 Minimum Cost to cut a board into squares/main.cpp
// Link: https://www.geeksforgeeks.org/problems/minimum-cost-to-cut-a-board-into-squares/1

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution
{
public:
    int minCost(int n, int m, vector<int> &x, vector<int> &y)
    {
        sort(x.begin(), x.end());
        sort(y.begin(), y.end());

        int hCount = 1, vCount = 1;
        int i = x.size() - 1, j = y.size() - 1;
        int totalCost = 0;

        while (i >= 0 && j >= 0)
        {
            if (x[i] >= y[j])
            {
                totalCost += x[i] * hCount;
                vCount++;
                i--;
            }
            else
            {
                totalCost += y[j] * vCount;
                hCount++;
                j--;
            }
        }

        while (i >= 0)
        {
            totalCost += x[i] * hCount;
            vCount++;
            i--;
        }

        while (j >= 0)
        {
            totalCost += y[j] * vCount;
            hCount++;
            j--;
        }

        return totalCost;
    }
};

int main()
{
    Solution sol;
    int n = 4, m = 6;
    vector<int> x = {2, 1, 3, 1, 4};
    vector<int> y = {4, 1, 2};
    int result = sol.minCost(n, m, x, y);
    cout << "Minimum cost to cut the board into squares: " << result << endl;

    n = 4, m = 4;
    x = {1, 1, 1};
    y = {1, 1, 1};
    result = sol.minCost(n, m, x, y);
    cout << "Minimum cost to cut the board into squares: " << result << endl;
    return 0;
}